ops/ccs-ops-misc/load_test/common/bfd_user_base.py

```diff
@@ -0,0 +1,183 @@
+'''Base class for Locust tests run against the FHIR endpoints.
+'''
+
+import json
+import logging
+import os
+
+from typing import Callable, Dict, List, Union
+from common import config, data, test_setup as setup, validation
+from common.url_path import create_url_path
+from locust import HttpUser
+
+import urllib3
+
+setup.set_locust_env(config.load())
+
+
+class BFDUserBase(HttpUser):
+    '''Base Class for Locust tests against BFD.
+
+    This class should automatically handle most of the common tasks that our load tests require.
+    '''
+
+    # Mark this class as abstract so Locust knows it doesn't contain Tasks
+    abstract = True
+
+    # The goals against which to measure these results. Note that they also include the Failsafe
+    # cutoff, which will default to the V2 cutoff time if not set.
+    VALIDATION_GOALS = None
+
+    # Do we terminate the tests when a test runs out of data and paginated URLs?
+    END_ON_NO_DATA = True
+
+    def __init__(self, *args, **kwargs):
+        HttpUser.__init__(self, *args, **kwargs)
+
+        # Load configuration needed for making requests to the FHIR server
+        self.client_cert = setup.get_client_cert()
+        self.server_public_key = setup.load_server_public_key()
+        setup.disable_no_cert_warnings(self.server_public_key, urllib3)
+        self.last_updated = data.get_last_updated()
+
+        # Initialize data / URL pools
+        self.bene_ids = []
+        self.mbis = []
+        self.contract_ids = []
+        self.url_pools = {}
+
+        self.logger = logging.getLogger()
+        self.has_reported_no_data = []
+
+
+    def on_start(self):
+        '''Run once when a BFDUser is initialized by Locust.
+        '''
+
+        # Is this either the first worker or the only worker?
+        worker_number = self.__get_worker_number()
+        if worker_number is None or str(worker_number) == '0':
+            # Adds a global failsafe check to ensure that if this test overwhelms
+            # the database, we bail out and stop hitting the server
+            if hasattr(self, 'VALIDATION_GOALS') and self.VALIDATION_GOALS:
+                validation.setup_failsafe_event(self.environment, self.VALIDATION_GOALS)
+            else:
+                validation.setup_failsafe_event(self.environment, validation.SLA_V2_BASELINE)
+
+
+    def on_stop(self):
+        '''Run tear-down tasks after the tests have completed.'''
+
+        # Report the various response time percentiles against the SLA
+        if hasattr(self, 'VALIDATION_GOALS') and self.VALIDATION_GOALS:
+            validation.check_sla_validation(self.environment, self.VALIDATION_GOALS)
+
+
+    def get_by_url(self, url: str, headers: Dict[str, str] = None,
+            name: str = ''):
```

Comment:
Curious about the explicit `None` here instead of an empty dictionary. That would save us the extra conditional one-liner below involving `safe_headers`, but... maybe I'm not seeing the whole picture.
```suggestion
    def get_by_url(self, url: str, headers: Dict[str, str] = {},
            name: str = ''):
```

Reply:
This one came from the linter, but after doing some research, it looks like it's a safety issue. If you use a default empty dictionary and then make changes to the dictionary, then subsequent invocations can/will receive the modified dictionary instead of a new empty one.

https://stackoverflow.com/questions/5587582/numpy-array-subclass-unexpedly-shares-attributes-across-instances

Reply:
No kidding? That's not very intuitive (to me) and probably pretty annoying to debug if encountered in the wild. Thanks for the explanation!